src/riak_cs_s3_auth.erl

```diff
@@ -130,43 +172,157 @@ calculate_signature(KeyData, RD) ->
+    case CalculatedSignature of
+        PresentedSignature -> ok;
+        _ -> {error, {unmatched_signature, PresentedSignature, CalculatedSignature}}
+    end.
+
+canonical_request_v4(AuthAttrs, Method, Path, Qs, AllHeaders) ->
+    CanonicalQs = canonicalize_qs(v4, Qs),
+    {"SignedHeaders", SignedHeaders} = lists:keyfind("SignedHeaders", 1, AuthAttrs),
+    CanonicalHeaders = canonical_headers_v4(AllHeaders,
+                                            string:tokens(SignedHeaders, [$;])),
+    {"x-amz-content-sha256", HashedPayload} =
+        lists:keyfind("x-amz-content-sha256", 1, AllHeaders),
+    CanonicalRequest = [atom_to_list(Method), $\n,
+                        strict_url_encode_for_path(Path), $\n,
+                        CanonicalQs, $\n,
+                        CanonicalHeaders, $\n,
+                        SignedHeaders, $\n,
+                        HashedPayload],
+    CanonicalRequest.
+
+canonical_headers_v4(AllHeaders, HeaderNames) ->
+    %% TODO:
+    %% - Assert `host' and all `x-amz-*' in AllHeaders are included in `HeaderNames'
+    %% - Assert `conetnt-type' is included in `HeaderNames' if it is included in
+    %%   `AllHeaders'
+    canonical_headers_v4(AllHeaders, HeaderNames, []).
+
+canonical_headers_v4(_AllHeaders, [], Acc) ->
+    Acc;
+canonical_headers_v4(AllHeaders, [HeaderName | Rest], Acc) ->
+    {HeaderName, Value} = lists:keyfind(HeaderName, 1, AllHeaders),
+    canonical_headers_v4(AllHeaders, Rest, [Acc | [HeaderName, $:, Value, $\n]]).
+
+string_to_sign_v4(AuthAttrs, AllHeaders, CanonicalRequest) ->
+    %% TODO: Expires?
+    TimeStamp = case lists:keyfind("x-amz-date", 1, AllHeaders) of
+                    false ->
+                        {"date", Date} = lists:keyfind("date", 1, AllHeaders),
+                        riak_cs_wm_utils:to_iso_8601(Date);
+                    {"x-amz-date", XAmzDate} ->
+                        XAmzDate
+                end,
+    {"Credential", Cred} = lists:keyfind("Credential", 1, AuthAttrs),
+    [_UserId, CredDate, AwsRegion, "s3" = AwsService, "aws4_request" = AwsRequest] =
+        string:tokens(Cred, [$/]),
+    %% TODO: Validate `CredDate' be within 7 days
+    Scope = [CredDate, $/, AwsRegion, $/, AwsService, $/, AwsRequest],
+    {["AWS4-HMAC-SHA256", $\n,
+      TimeStamp, $\n,
+      Scope, $\n,
+      hex_sha256hash(CanonicalRequest)],
+     {CredDate, AwsRegion, AwsService, AwsRequest}}.
+
+calculate_signature_v4(SecretAccessKey,
+                       {CredDate, AwsRegion, AwsService, AwsRequest}, StringToSign) ->
+    DateKey              = hmac_sha256(["AWS4", SecretAccessKey], CredDate),
+    DateRegionKey        = hmac_sha256(DateKey, AwsRegion),
+    DateRegionServiceKey = hmac_sha256(DateRegionKey, AwsService),
+    SigningKey           = hmac_sha256(DateRegionServiceKey, AwsRequest),
+    mochihex:to_hex(hmac_sha256(SigningKey, StringToSign)).
+
+hmac_sha256(Key, Data) ->
+    %% TODO: R15* compatibility?
```

Comment:
Looks like this function could be moved to `riak_cs_utils` and put it under ifdefs. Or even they could be wrapped at webmachine? We can do it when we need it, later.